File: extensions/deepstream/sidecar/protocol.py
"""JSON Lines wire protocol for the DeepStream sidecar.

Mirror of the Rust enum ``protocol.rs``. Tag-keyed discriminated unions
(``{"type": "<snake_case_tag>", ...fields}``) — matches the Rust
``#[serde(tag = "type", rename_all = "snake_case")]`` exactly.

Wire format (single JSON object per line, terminated by ``\n``):

- Control messages (host -> sidecar, read from stdin):

    * ``hello``            { rtsp_port, snapshot_port, log_level,
                             models_dir, max_streams, snapshot_bind_addr }
    * ``add_stream``       { id, config: <StreamConfig dict> }
    * ``remove_stream``    { id, stream_id, graceful_secs }
    * ``update_analytics`` { id, stream_id, line_crossing, roi }
    * ``set_threshold``    { id, stream_id, conf, iou }
    * ``list_state``       { id }
    * ``health_check``     { ts }
    * ``shutdown``         { graceful_secs }

- Sidecar events (sidecar -> host, written to stdout):

    * ``ready``             { ds_ver, pyds_ver, protocol_ver, gpu_info }
    * ``hello_ack``         { max_streams, rtsp_url_prefix, models_loaded }
    * ``stream_added``      { id, stream_id, rtsp_url }
    * ``stream_removed``    { id, stream_id }
    * ``stream_error``      { stream_id, code, message, id? }
    * ``detection``         { stream_id, ts, frame_id, objects: [...] }
    * ``line_cross``        { stream_id, ts, line_id, track_id, class, direction }
    * ``roi_intrusion``     { stream_id, ts, roi_id, track_id, class, mode }
    * ``analytics_snapshot``{ stream_id, ts, snapshot }
    * ``stats``             { ts, global_fps, gpu_utilization_percent,
                              gpu_memory_used_mb, per_stream: [...] }
    * ``pong``              { ts }
    * ``error_response``    { id, code, message }
    * ``bye``               { reason, exit_code }

This module uses ONLY the Python stdlib so it can be imported on macOS
for unit tests of the wire format. Anything that touches ``pyds`` /
``Gst`` / ``GLib`` lives in other modules.

``__main__`` smoke test (Task 8.2 Step 2): reads stdin line-by-line,
parses each line as a ControlMessage, and echoes any ``health_check``
back as a ``pong`` event on stdout. Empty stdin -> no output, exit 0.
"""

from __future__ import annotations

import dataclasses
import json
import sys
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Type, Union

# --- Constants (mirror protocol.rs) ----------------------------------------

PROTOCOL_VERSION: int = 1
MAX_LINE_BYTES: int = 4 * 1024 * 1024  # 4 MiB, matches Rust.


class ProtocolError(Exception):
    """Raised when a line cannot be serialized / deserialized."""


class LineTooLongError(ProtocolError):
    """A line exceeded MAX_LINE_BYTES."""


# --- Helpers ---------------------------------------------------------------


def _snake(s: str) -> str:
    """CamelCase -> snake_case (only for error messages / debug)."""
    out = []
    for i, ch in enumerate(s):
        if ch.isupper() and i > 0:
            out.append("_")
        out.append(ch.lower())
    return "".join(out)


def serialize(obj: Any) -> bytes:
    """Serialize ``obj`` to a single JSON line terminated by ``\\n``.

    ``obj`` must be a dataclass instance whose ``to_dict()`` (or whose
    ``dataclasses.asdict`` output) yields the canonical wire shape,
    including a ``"type"`` discriminator. Raises :class:`LineTooLongError`
    if the resulting line exceeds :data:`MAX_LINE_BYTES`.
    """
    payload = to_wire_dict(obj)
    line = json.dumps(payload, separators=(",", ":"))
    if len(line) > MAX_LINE_BYTES:
        raise LineTooLongError(
            f"line exceeds {MAX_LINE_BYTES} bytes ({len(line)})"
        )
    return (line + "\n").encode("utf-8")


def to_wire_dict(obj: Any) -> Dict[str, Any]:
    """Convert a dataclass instance to its wire-format dict.

    - If the class has an explicit ``to_wire_dict`` method, defer to it
      (covers nested structs that don't carry a ``type`` tag, and event
      variants that need special handling like Stats).
    - Otherwise the dataclass is a "plain" tagged event: emit
      ``"type": TYPE_TAG`` plus all dataclass fields, omitting fields in
      ``OPTIONAL_OMIT_NONE`` whose value is None.
    """
    if hasattr(obj, "to_wire_dict") and callable(obj.to_wire_dict):
        return obj.to_wire_dict()
    if dataclasses.is_dataclass(obj):
        d: Dict[str, Any] = {}
        tag = getattr(obj, "TYPE_TAG", None)
        if tag is not None:
            d["type"] = tag
        omit_none = getattr(obj, "OPTIONAL_OMIT_NONE", frozenset())
        for f in dataclasses.fields(obj):
            if f.name in {"TYPE_TAG", "OPTIONAL_OMIT_NONE"}:
                continue
            val = getattr(obj, f.name)
            if val is None and f.name in omit_none:
                continue
            d[f.name] = _to_primitive(val)
        return d
    raise TypeError(f"cannot serialize {type(obj).__name__}")


def _to_primitive(val: Any) -> Any:
    """Recursively convert dataclass / tuple / list / dict to JSON-safe."""
    if hasattr(val, "to_wire_dict") and callable(val.to_wire_dict):
        return val.to_wire_dict()
    if dataclasses.is_dataclass(val) and not isinstance(val, type):
        return to_wire_dict(val)
    if isinstance(val, (list, tuple)):
        return [_to_primitive(v) for v in val]
    if isinstance(val, dict):
        return {k: _to_primitive(v) for k, v in val.items()}
    return val


def deserialize_line(line: str) -> Dict[str, Any]:
    """Parse one JSON line into a raw dict (caller dispatches on ``type``).

    The dict is NOT validated against the protocol — it is the raw
    payload as returned by :func:`json.loads`. The typed accessors
    (:class:`Hello`, :func:`parse_control_message`, etc.) provide the
    type-checked view.
    """
    if len(line) > MAX_LINE_BYTES:
        raise LineTooLongError(
            f"line exceeds {MAX_LINE_BYTES} bytes ({len(line)})"
        )
    try:
        return json.loads(line)
    except json.JSONDecodeError as e:
        raise ProtocolError(f"invalid JSON: {e}") from e


# --- GPU info / detection objects / stream stats ---------------------------


@dataclass
class GpuInfo:
    """Nested struct — NOT a tagged event variant, no ``type`` field."""
    name: str
    mem_mb: int

    def to_wire_dict(self) -> Dict[str, Any]:
        return {"name": self.name, "mem_mb": self.mem_mb}


@dataclass
class DetectionObject:
    """Nested struct — NOT a tagged event variant.

    ``bbox`` is ``[left, top, right, bottom]``.
    """
    class_: int
    conf: float
    bbox: List[float]
    track_id: Optional[int] = None

    def to_wire_dict(self) -> Dict[str, Any]:
        # ``class`` is a reserved word in Python — we expose ``class_`` but
        # the wire key is ``class`` (matches Rust field ``class: u32``).
        d: Dict[str, Any] = {
            "class": self.class_,
            "conf": self.conf,
            "bbox": list(self.bbox),
        }
        if self.track_id is not None:
            d["track_id"] = self.track_id
        return d


@dataclass
class StreamStat:
    """Nested struct — NOT a tagged event variant."""
    stream_id: str
    fps: float
    latency_ms: float
    frame_count: int
    object_count: int
    status: str

    def to_wire_dict(self) -> Dict[str, Any]:
        return {
            "stream_id": self.stream_id,
            "fps": self.fps,
            "latency_ms": self.latency_ms,
            "frame_count": self.frame_count,
            "object_count": self.object_count,
            "status": self.status,
        }


# --- Stats payload ---------------------------------------------------------
#
# The Rust enum has ``SidecarEvent::Stats(Stats)`` — a tuple variant.
# Tuple variants do NOT add a wrapping field; their inner fields are
# inlined directly into the event object alongside the tag. We mirror
# that here by making ``Stats`` carry the tag itself.

@dataclass
class Stats:
    ts: int
    global_fps: float
    gpu_utilization_percent: float
    gpu_memory_used_mb: float
    per_stream: List[StreamStat] = field(default_factory=list)

    TYPE_TAG: str = field(init=False, default="stats", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)

    def to_wire_dict(self) -> Dict[str, Any]:
        return {
            "type": "stats",
            "ts": self.ts,
            "global_fps": self.global_fps,
            "gpu_utilization_percent": self.gpu_utilization_percent,
            "gpu_memory_used_mb": self.gpu_memory_used_mb,
            "per_stream": [to_wire_dict(s) for s in self.per_stream],
        }


# --- Control messages (host -> sidecar) ------------------------------------


@dataclass
class Hello:
    rtsp_port: int
    snapshot_port: int
    log_level: str
    models_dir: str
    max_streams: int
    snapshot_bind_addr: str

    TYPE_TAG: str = field(init=False, default="hello", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class AddStream:
    """NOTE: there is NO top-level ``stream_id`` field — it lives inside ``config``."""
    id: str
    config: Dict[str, Any]  # raw StreamConfig JSON — parsed by config.py.

    TYPE_TAG: str = field(init=False, default="add_stream", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class RemoveStream:
    id: str
    stream_id: str
    graceful_secs: int

    TYPE_TAG: str = field(init=False, default="remove_stream", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class UpdateAnalytics:
    id: str
    stream_id: str
    line_crossing: Dict[str, Any]
    roi: Dict[str, Any]

    TYPE_TAG: str = field(init=False, default="update_analytics", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class SetThreshold:
    id: str
    stream_id: str
    conf: float
    iou: float

    TYPE_TAG: str = field(init=False, default="set_threshold", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class ListState:
    id: str

    TYPE_TAG: str = field(init=False, default="list_state", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class HealthCheck:
    ts: int

    TYPE_TAG: str = field(init=False, default="health_check", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class Shutdown:
    graceful_secs: int

    TYPE_TAG: str = field(init=False, default="shutdown", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


_CONTROL_REGISTRY: Dict[str, Type] = {
    "hello": Hello,
    "add_stream": AddStream,
    "remove_stream": RemoveStream,
    "update_analytics": UpdateAnalytics,
    "set_threshold": SetThreshold,
    "list_state": ListState,
    "health_check": HealthCheck,
    "shutdown": Shutdown,
}

ControlMessage = Union[
    Hello, AddStream, RemoveStream, UpdateAnalytics,
    SetThreshold, ListState, HealthCheck, Shutdown,
]


def parse_control_message(d: Dict[str, Any]) -> ControlMessage:
    """Type-check a raw dict into a :data:`ControlMessage` dataclass.

    Raises :class:`ProtocolError` if ``type`` is missing or unknown.
    Extra keys are ignored (forward-compat with new protocol fields).
    """
    tag = d.get("type")
    if not isinstance(tag, str):
        raise ProtocolError(f"missing/invalid 'type' tag: {tag!r}")
    cls = _CONTROL_REGISTRY.get(tag)
    if cls is None:
        raise ProtocolError(f"unknown control message type: {tag!r}")
    return _build_dataclass(cls, d)  # type: ignore[return-value]


def _build_dataclass(cls: Type, d: Dict[str, Any]) -> Any:
    """Construct ``cls`` from dict ``d`` using only known fields."""
    field_names = {f.name for f in dataclasses.fields(cls)
                   if f.name not in {"TYPE_TAG", "OPTIONAL_OMIT_NONE"}}
    kwargs: Dict[str, Any] = {}
    for fn in field_names:
        if fn in d:
            kwargs[fn] = d[fn]
    try:
        return cls(**kwargs)
    except TypeError as e:
        raise ProtocolError(f"cannot build {cls.__name__}: {e}") from e


# --- Sidecar events (sidecar -> host) --------------------------------------


@dataclass
class Ready:
    ds_ver: str
    pyds_ver: str
    protocol_ver: int
    gpu_info: GpuInfo

    TYPE_TAG: str = field(init=False, default="ready", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)

    def to_wire_dict(self) -> Dict[str, Any]:
        return {
            "type": "ready",
            "ds_ver": self.ds_ver,
            "pyds_ver": self.pyds_ver,
            "protocol_ver": self.protocol_ver,
            "gpu_info": to_wire_dict(self.gpu_info),
        }


@dataclass
class HelloAck:
    max_streams: int
    rtsp_url_prefix: str
    models_loaded: List[str] = field(default_factory=list)

    TYPE_TAG: str = field(init=False, default="hello_ack", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)

    def to_wire_dict(self) -> Dict[str, Any]:
        return {
            "type": "hello_ack",
            "max_streams": self.max_streams,
            "rtsp_url_prefix": self.rtsp_url_prefix,
            "models_loaded": list(self.models_loaded),
        }


@dataclass
class StreamAdded:
    id: str
    stream_id: str
    rtsp_url: str
    snapshot_token: str = ""

    TYPE_TAG: str = field(init=False, default="stream_added", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class StreamRemoved:
    id: str
    stream_id: str

    TYPE_TAG: str = field(init=False, default="stream_removed", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class StreamError:
    stream_id: str
    code: str
    message: str
    id: Optional[str] = None

    TYPE_TAG: str = field(init=False, default="stream_error", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(
        init=False, default=frozenset({"id"}), repr=False
    )

    def to_wire_dict(self) -> Dict[str, Any]:
        d: Dict[str, Any] = {
            "type": "stream_error",
            "stream_id": self.stream_id,
            "code": self.code,
            "message": self.message,
        }
        if self.id is not None:
            d["id"] = self.id
        return d


@dataclass
class Detection:
    stream_id: str
    ts: int
    frame_id: int
    objects: List[DetectionObject] = field(default_factory=list)

    TYPE_TAG: str = field(init=False, default="detection", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)

    def to_wire_dict(self) -> Dict[str, Any]:
        return {
            "type": "detection",
            "stream_id": self.stream_id,
            "ts": self.ts,
            "frame_id": self.frame_id,
            "objects": [o.to_wire_dict() for o in self.objects],
        }


@dataclass
class LineCross:
    stream_id: str
    ts: int
    line_id: str
    track_id: int
    class_: int
    direction: str

    TYPE_TAG: str = field(init=False, default="line_cross", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)

    def to_wire_dict(self) -> Dict[str, Any]:
        return {
            "type": "line_cross",
            "stream_id": self.stream_id,
            "ts": self.ts,
            "line_id": self.line_id,
            "track_id": self.track_id,
            "class": self.class_,
            "direction": self.direction,
        }


@dataclass
class ROIIntrusion:
    stream_id: str
    ts: int
    roi_id: str
    track_id: int
    class_: int
    mode: str

    TYPE_TAG: str = field(init=False, default="roi_intrusion", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)

    def to_wire_dict(self) -> Dict[str, Any]:
        return {
            "type": "roi_intrusion",
            "stream_id": self.stream_id,
            "ts": self.ts,
            "roi_id": self.roi_id,
            "track_id": self.track_id,
            "class": self.class_,
            "mode": self.mode,
        }


@dataclass
class AnalyticsSnapshot:
    stream_id: str
    ts: int
    snapshot: Dict[str, Any]

    TYPE_TAG: str = field(init=False, default="analytics_snapshot", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class Pong:
    ts: int

    TYPE_TAG: str = field(init=False, default="pong", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class ErrorResponse:
    id: str
    code: str
    message: str

    TYPE_TAG: str = field(init=False, default="error_response", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


@dataclass
class Bye:
    reason: str
    exit_code: int

    TYPE_TAG: str = field(init=False, default="bye", repr=False)
    OPTIONAL_OMIT_NONE: frozenset = field(init=False, default=frozenset(), repr=False)


SidecarEvent = Union[
    Ready, HelloAck, StreamAdded, StreamRemoved, StreamError,
    Detection, LineCross, ROIIntrusion, AnalyticsSnapshot,
    Stats, Pong, ErrorResponse, Bye,
]


# --- Asyncio line reader ---------------------------------------------------
#
# Optional helper for ``asyncio``-based runners. Kept here so that the
# protocol module is the one-stop shop for wire I/O. Pure-stdlib.

async def read_message_async(reader: "Any") -> Dict[str, Any]:
    """Read one JSON line from an asyncio ``StreamReader``.

    Raises :class:`LineTooLongError` if the next line (excluding newline)
    exceeds :data:`MAX_LINE_BYTES`, matching the Rust behavior.
    """

    buf = bytearray()
    while True:
        chunk = await reader.read(4096)
        if not chunk:
            if not buf:
                raise EOFError("stream closed before any data")
            break
        nl = chunk.find(b"\n")
        if nl >= 0:
            buf.extend(chunk[:nl])
            # NOTE: any bytes after the newline are silently dropped.
            # Callers should use a buffered reader (e.g. asyncio's
            # StreamReader) that handles framing itself.
            break
        buf.extend(chunk)
        if len(buf) > MAX_LINE_BYTES:
            raise LineTooLongError(f"line exceeds {MAX_LINE_BYTES} bytes")
    if len(buf) > MAX_LINE_BYTES:
        raise LineTooLongError(f"line exceeds {MAX_LINE_BYTES} bytes")
    line = buf.decode("utf-8")
    return deserialize_line(line)


# --- Smoke test (Task 8.2 Step 2) ------------------------------------------
#
# Reads ControlMessage lines from stdin, echoes any ``health_check``
# back as a ``pong`` event on stdout. Empty stdin -> no output, exit 0.

def _main() -> int:

    for line in sys.stdin:
        line = line.rstrip("\n")
        if not line:
            continue
        try:
            msg = parse_control_message(deserialize_line(line))
        except ProtocolError as e:
            err = ErrorResponse(id="?", code="parse_error", message=str(e))
            sys.stdout.buffer.write(serialize(err))
            sys.stdout.buffer.flush()
            continue
        if isinstance(msg, HealthCheck):
            pong = Pong(ts=msg.ts)
            sys.stdout.buffer.write(serialize(pong))
            sys.stdout.buffer.flush()
    return 0


if __name__ == "__main__":
    raise SystemExit(_main())

File: extensions/deepstream/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde::de::DeserializeOwned;
use std::io;
use tokio::io::{AsyncRead, AsyncWrite, AsyncWriteExt};

pub const MAX_LINE_BYTES: usize = 4 * 1024 * 1024;

#[derive(Debug, thiserror::Error)]
pub enum ProtocolError {
    #[error("line exceeds {MAX_LINE_BYTES} bytes")]
    LineTooLong,
    #[error("invalid utf-8 in line")]
    InvalidUtf8,
    #[error("io: {0}")]
    Io(#[from] io::Error),
    #[error("json: {0}")]
    Json(#[from] serde_json::Error),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ControlMessage {
    Hello {
        rtsp_port: u16,
        snapshot_port: u16,
        log_level: String,
        models_dir: String,
        max_streams: u32,
        snapshot_bind_addr: String,
    },
    AddStream { id: String, config: serde_json::Value },
    RemoveStream { id: String, stream_id: String, graceful_secs: u32 },
    UpdateAnalytics { id: String, stream_id: String, line_crossing: serde_json::Value, roi: serde_json::Value },
    SetThreshold { id: String, stream_id: String, conf: f32, iou: f32 },
    ListState { id: String },
    HealthCheck { ts: i64 },
    Shutdown { graceful_secs: u32 },
}

pub fn serialize_line(msg: &ControlMessage) -> Result<String, ProtocolError> {
    let mut s = serde_json::to_string(msg)?;
    if s.len() > MAX_LINE_BYTES { return Err(ProtocolError::LineTooLong); }
    s.push('\n');
    Ok(s)
}

pub fn deserialize_line(line: &str) -> Result<ControlMessage, ProtocolError> {
    if line.len() > MAX_LINE_BYTES { return Err(ProtocolError::LineTooLong); }
    Ok(serde_json::from_str(line)?)
}

pub async fn write_message<W: AsyncWrite + Unpin>(
    w: &mut W,
    msg: &ControlMessage,
) -> Result<(), ProtocolError> {
    let mut line = serde_json::to_string(msg)?;
    if line.len() > MAX_LINE_BYTES {
        return Err(ProtocolError::LineTooLong);
    }
    line.push('\n');
    w.write_all(line.as_bytes()).await.map_err(ProtocolError::Io)?;
    Ok(())
}

pub async fn read_message<R: AsyncRead + Unpin, T: DeserializeOwned>(
    reader: &mut tokio::io::BufReader<R>,
) -> Result<T, ProtocolError> {
    use tokio::io::AsyncBufReadExt;
    loop {
        let mut buf: Vec<u8> = Vec::with_capacity(4096);
        // read_until properly preserves BufReader internal state across
        // calls (unlike manual read() which loses bytes after the newline).
        let n = reader.read_until(b'\n', &mut buf).await.map_err(ProtocolError::Io)?;
        if n == 0 {
            // EOF — underlying reader closed. This is a real end-of-stream,
            // not a blank line in the middle of the banner.
            return Err(ProtocolError::Io(std::io::ErrorKind::UnexpectedEof.into()));
        }
        if buf.len() > MAX_LINE_BYTES {
            return Err(ProtocolError::LineTooLong);
        }

        // Strip the trailing newline (read_until includes it).
        if buf.last() == Some(&b'\n') {
            buf.pop();
            if buf.last() == Some(&b'\r') {
                buf.pop();
            }
        }

        // Skip empty lines and non-JSON lines. The DeepStream sidecar runs
        // inside the NVIDIA NGC container whose entrypoint prints a CUDA
        // banner (`\n==========\n== CUDA ==\n==========\n...`) before
        // deepstream_runner.py emits its first JSONL frame. Any of those
        // banner lines would fail serde_json::from_str and previously
        // terminated the reader_loop, causing the extension to fail
        // "sidecar stdout closed before Ready" on every remote-mode spawn.
        // Real JSONL messages start with `{`; banner lines never do.
        let first_non_ws = buf.iter().copied().find(|&b| !b.is_ascii_whitespace());
        if first_non_ws != Some(b'{') {
            let preview = std::str::from_utf8(&buf)
                .unwrap_or("<non-utf8>")
                .chars()
                .take(80)
                .collect::<String>();
            eprintln!(
                "[deepstream] skipping non-JSON preamble line ({}B): {:?}",
                buf.len(),
                preview
            );
            continue;
        }

        let s = std::str::from_utf8(&buf).map_err(|_| ProtocolError::InvalidUtf8)?;
        return serde_json::from_str(s).map_err(Into::into);
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SidecarEvent {
    Ready {
        ds_ver: String,
        pyds_ver: String,
        protocol_ver: u32,
        gpu_info: GpuInfo,
    },
    HelloAck {
        max_streams: u32,
        rtsp_url_prefix: String,
        models_loaded: Vec<String>,
    },
    StreamAdded {
        id: String,
        stream_id: String,
        rtsp_url: String,
        snapshot_token: String,
    },
    StreamRemoved {
        id: String,
        stream_id: String,
    },
    StreamError {
        stream_id: String,
        code: String,
        message: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        id: Option<String>,
    },
    Detection {
        stream_id: String,
        ts: i64,
        frame_id: u64,
        objects: Vec<DetectionObject>,
    },
    LineCross {
        stream_id: String,
        ts: i64,
        line_id: String,
        track_id: u32,
        class: u32,
        direction: String,
    },
    ROIIntrusion {
        stream_id: String,
        ts: i64,
        roi_id: String,
        track_id: u32,
        class: u32,
        mode: String,
    },
    AnalyticsSnapshot {
        stream_id: String,
        ts: i64,
        snapshot: serde_json::Value,
    },
    Stats(Stats),
    Pong { ts: i64 },
    ErrorResponse {
        id: String,
        code: String,
        message: String,
    },
    Bye {
        reason: String,
        exit_code: i32,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GpuInfo {
    pub name: String,
    pub mem_mb: u32,
}

/// Handshake ack fields extracted from [`SidecarEvent::HelloAck`] so callers
/// (e.g. `perform_handshake`) can consume them without pattern-matching the
/// whole enum. Mirrors the variant's payload.
#[derive(Debug, Clone)]
pub struct HelloAck {
    pub max_streams: u32,
    pub rtsp_url_prefix: String,
    pub models_loaded: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DetectionObject {
    pub class: u32,
    pub conf: f32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub track_id: Option<u32>,
    pub bbox: [f32; 4], // left, top, right, bottom — Python decides normalization
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StreamStat {
    pub stream_id: String,
    pub fps: f32,
    pub latency_ms: f32,
    pub frame_count: u64,
    pub object_count: u32,
    pub status: String,
}

/// Sidecar-wide stats snapshot. Carried by [`SidecarEvent::Stats`].
///
/// `global_fps` is the aggregate throughput across all streams; it surfaces
/// as the `total_throughput_fps` metric on the host side (the rename reflects
/// the metric's display semantics, not a semantic change to the data).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Stats {
    pub ts: i64,
    pub global_fps: f32,
    pub gpu_utilization_percent: f32,
    pub gpu_memory_used_mb: f32,
    pub per_stream: Vec<StreamStat>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn serialize_add_stream_minimal() {
        let msg = ControlMessage::AddStream {
            id: "r1".into(),
            config: serde_json::json!({"source":{"type":"rtsp","url":"rtsp://x"},"model":"yolov8n-coco"}),
        };
        let line = serialize_line(&msg).unwrap();
        assert!(line.matches('\n').count() == 1 && line.ends_with('\n'));  // single JSON line + newline terminator
        let parsed: ControlMessage = serde_json::from_str(line.trim()).unwrap();
        match parsed { ControlMessage::AddStream { id, .. } => assert_eq!(id, "r1"), _ => panic!() }
    }

    #[test]
    fn reject_line_over_4mb() {
        let huge = "x".repeat(4 * 1024 * 1024 + 1);
        let err = deserialize_line(&huge).unwrap_err();
        assert!(matches!(err, ProtocolError::LineTooLong));
    }

    #[test]
    fn parse_ready_event() {
        let line = r#"{"type":"ready","ds_ver":"7.1.0","pyds_ver":"1.1.1","protocol_ver":1,"gpu_info":{"name":"Orin NX","mem_mb":8192}}"#;
        let ev: SidecarEvent = serde_json::from_str(line).unwrap();
        match ev {
            SidecarEvent::Ready { ds_ver, protocol_ver, .. } => {
                assert_eq!(ds_ver, "7.1.0");
                assert_eq!(protocol_ver, 1);
            }
            _ => panic!(),
        }
    }

    #[tokio::test]
    async fn round_trip_three_messages_via_duplex() {
        use tokio::io::{AsyncReadExt, AsyncWriteExt};
        let (mut tx, mut rx) = tokio::io::duplex(8 * 1024);
        let msgs = vec![
            ControlMessage::AddStream {
                id: "r1".into(),
                config: serde_json::json!({"source":{"type":"rtsp","url":"rtsp://a"}}),
            },
            ControlMessage::RemoveStream {
                id: "r2".into(),
                stream_id: "r1".into(),
                graceful_secs: 2,
            },
            ControlMessage::HealthCheck { ts: 12345 },
        ];

        let writer = tokio::spawn(async move {
            for m in &msgs {
                write_message(&mut tx, m).await.unwrap();
            }
            tx.shutdown().await.unwrap();
        });

        let mut buf = Vec::new();
        rx.read_to_end(&mut buf).await.unwrap();
        writer.await.unwrap();

        let text = String::from_utf8(buf).unwrap();
        let lines: Vec<&str> = text.lines().filter(|l| !l.is_empty()).collect();
        assert_eq!(lines.len(), 3);
        let parsed: Vec<ControlMessage> = lines.iter()
            .map(|l| serde_json::from_str(l).unwrap())
            .collect();
        assert!(matches!(parsed[0], ControlMessage::AddStream { .. }));
        assert!(matches!(parsed[1], ControlMessage::RemoveStream { .. }));
        assert!(matches!(parsed[2], ControlMessage::HealthCheck { .. }));
    }

    #[tokio::test]
    async fn write_message_appends_newline() {
        use tokio::io::AsyncReadExt;
        let (mut tx, mut rx) = tokio::io::duplex(1024);
        write_message(&mut tx, &ControlMessage::HealthCheck { ts: 1 }).await.unwrap();
        tx.shutdown().await.unwrap();
        let mut buf = Vec::new();
        rx.read_to_end(&mut buf).await.unwrap();
        let s = String::from_utf8(buf).unwrap();
        assert!(s.ends_with('\n'));
        assert_eq!(s.matches('\n').count(), 1);
    }

    #[tokio::test]
    async fn read_message_round_trip_sidecar_event() {
        use tokio::io::{AsyncWriteExt, BufReader};
        let (mut tx, rx) = tokio::io::duplex(8 * 1024);
        let event_line = r#"{"type":"pong","ts":999}"#;
        tx.write_all(event_line.as_bytes()).await.unwrap();
        tx.write_all(b"\n").await.unwrap();

        let mut reader = BufReader::new(rx);
        let ev = read_message::<_, SidecarEvent>(&mut reader).await.unwrap();
        match ev {
            SidecarEvent::Pong { ts } => assert_eq!(ts, 999),
            _ => panic!("expected Pong"),
        }
    }

    #[tokio::test]
    async fn read_message_rejects_line_over_4mb() {
        use tokio::io::{AsyncWriteExt, BufReader};
        let (mut tx, rx) = tokio::io::duplex(8 * 1024 * 1024);
        let huge = format!("{}\n", "x".repeat(4 * 1024 * 1024 + 1));
        tx.write_all(huge.as_bytes()).await.unwrap();
        tx.shutdown().await.unwrap();

        let mut reader = BufReader::new(rx);
        let err = read_message::<_, SidecarEvent>(&mut reader).await.unwrap_err();
        assert!(matches!(err, ProtocolError::LineTooLong), "got {:?}", err);
    }

    /// Reproduces the bridge-mode bug where the NVIDIA NGC container entrypoint
    /// prints a CUDA banner before the sidecar emits its first JSONL `ready`.
    /// The original read_message would parse the first banner line, fail
    /// serde_json::from_str, and reader_loop would terminate — surfacing as
    /// "sidecar stdout closed before Ready" in perform_handshake. The fix
    /// skips any line whose first non-whitespace byte isn't `{`.
    #[tokio::test]
    async fn read_message_skips_ngc_container_banner() {
        use tokio::io::{AsyncWriteExt, BufReader};
        let banner = b"\n==========\n== CUDA ==\n==========\n\nCUDA Version 12.6.11\n\nContainer image Copyright (c) 2016-2023, NVIDIA CORPORATION & AFFILIATES.\n";
        let ready = br#"{"type":"ready","ds_ver":"7.1","pyds_ver":"unknown","protocol_ver":1,"gpu_info":{"name":"X","mem_mb":1024}}
"#;
        let (mut tx, rx) = tokio::io::duplex(64 * 1024);
        tx.write_all(banner).await.unwrap();
        tx.write_all(ready).await.unwrap();
        tx.shutdown().await.unwrap();

        let mut reader = BufReader::new(rx);
        let ev = read_message::<_, SidecarEvent>(&mut reader).await.expect("must skip banner and parse ready");
        match ev {
            SidecarEvent::Ready { ds_ver, .. } => assert_eq!(ds_ver, "7.1"),
            other => panic!("expected Ready, got {:?}", other),
        }
    }

    /// Plain empty lines (just `\n`) should be skipped, not treated as EOF.
    #[tokio::test]
    async fn read_message_skips_blank_lines() {
        use tokio::io::{AsyncWriteExt, BufReader};
        let (mut tx, rx) = tokio::io::duplex(8 * 1024);
        tx.write_all(b"\n\n\n").await.unwrap();
        tx.write_all(br#"{"type":"pong","ts":42}"#).await.unwrap();
        tx.write_all(b"\n").await.unwrap();
        tx.shutdown().await.unwrap();

        let mut reader = BufReader::new(rx);
        let ev = read_message::<_, SidecarEvent>(&mut reader).await.expect("must skip blanks");
        match ev {
            SidecarEvent::Pong { ts } => assert_eq!(ts, 42),
            other => panic!("expected Pong, got {:?}", other),
        }
    }
}
